data_safe_haven/infrastructure/project_manager.py

```diff
@@ -105,10 +113,23 @@ def project_settings(self) -> automation.ProjectSettings:
     def stack_settings(self) -> automation.StackSettings:
         return automation.StackSettings(
             config=self.pulumi_project.stack_config,
-            encrypted_key=self.pulumi_project.encrypted_key,
+            encrypted_key=self.pulumi_config.encrypted_key,
             secrets_provider=self.context.pulumi_secrets_provider_url,
         )
 
+    @property
+    def pulumi_project(self) -> DSHPulumiProject:
+        if not self._pulumi_project:
+            # Create DSH Pulumi Project if it does not exist, otherwise use existing
+            if self.create_project:
+                self._pulumi_project = self.pulumi_config.create_or_select_project(
+                    self.pulumi_project_name
+                )
+            else:
+                self._pulumi_project = self.pulumi_config[self.pulumi_project_name]
+
```

Comment:
We should catch the KeyError if `self.pulumi_project_name` is not an existing project an raise an informative exception.

Reply:
I thought this would handle that

https://github.com/alan-turing-institute/data-safe-haven/blob/063131becebb476c17bc08ecb08e05c32c620663/data_safe_haven/config/pulumi.py#L17-L26

Could we do something better here?
Like `"No SHM/SRE called {self.pulumi_project_name} is defined"`?

Reply:
Sorry - didn't notice that you'd overwritten `__getitem__`. In which case, yes, this should catch it.

Reply:
However, I think it's good practice to catch all exceptions that we think might happen and reraise them as some kind of `DataSafeHavenError` so that we can do an overall `try: ... catch DataSafeHavenError as exc: ...` in the main cli loop.